### DISCUSSION TITLE
Understanding $store

### DISCUSSION THREAD

**Discussion by ench0** [alpinejs/alpine#1504] [1. Help] [answered]
I am trying to decouple my data from the frontend but failing to understand how the store works in v3. So far I was using Alpine as I would do with jQuery - for basic frontend manipulation, but I want to take it a bit further and replace my React/Vue/Svelte stack. This is my store called 'tick':

```js
Alpine.store('tick', {
    count: 0,

    init() {
        const int1 = setInterval(() => {
            console.log('tick count:', this.count)
            this.get()
        }, 5000)
    },

    get() {
        return {
            count: this.count++,
        }
    },
})
```

it has 'count' integer value, 'init' (called automatically in v3) and 'get' function.

If I have this in my frontend, in this case nothing happens (the div shows initial value of 0) but the console log prints the proper value:

```html
<section x-data="$store.tick.get()">
    <div x-text="count"></div>
</section>
```

If I scope my store differently as per below, the div updates, but runs twice on each tick (shown both in the div and console):

```html
<section x-data="$store.tick">
    <div x-text="get().count"></div>
</section>
```
or
```html
<section x-data="$store.tick">
    <div x-text="count"></div>
</section>
```

Of course, this is but a basic example, while my real-world scenario would include much more code, like fetching remote images and such. Please help me understand the logic as the current docs do not deal with any advanced usage. I tried various 'magics' like '$watch' and such but to no avail. Should I look for more advanced state handling? Thanks in advance.

**Comment by ench0** [comment, marked as answer]
Follow-up: seems to be an issue with 'init' keyword. If I replace it with anything else, it works as expected. It just needs to be referenced from the root element:

```html
<section x-data="$store.tick" x-init="$store.tick.start()">
    <div x-text="count"></div>
</section>
```

While my store is this:

```js
Alpine.store('tick', {
    count: 0,

    start() {
        const int1 = setInterval(async () => {
            this.get()
            console.log('tick count:', this.count)
        }, 5000)
    },

    get() {
        return {
            count: this.count++,
        }
    },
})
```

...notice ie 'start' instead of 'init'. Anyway, if somebody has more insight or I'm doing something very nasty in the code above, please advise :smiley:

**Reply by danddanddand** [reply]
I've never seen a $store used as state for the component in way you're doing it. $store should be thought of as it's own global object that is reactive, so x-data="$store.anything" would be very abnormal usage. You probably just want

```
<section x-data>
    <div x-text="$store.tick.count"></div>
</section>

```

And the actual store either gets initialized in your bundle before Alpine.start() or with 'alpine:initializing' event.

**Reply by ench0** [reply]
Unfortunately my use case is not that simple. My store contains vars, computed vars based on other vars. On top of all I do some fetching of external api. I actually have it working just fine with that init substitute.

I realise Alpine might not be the best choice for that but I really enjoy using it.

**Reply by danddanddand** [reply]
You can have large store with any amount of computed values and any amount of logic. I just think you'd benefit from thinking of component state and global store state separately. If you had empty x-data and count referencing $store.tick.count and drop the getter to increment you get the same thing with less moving parts. You're essentially using the store as a factory method to create component state in your updated example and it might be unwieldy pattern so if you stick with this I'd use caution.

**Reply by ench0** [reply]
Thanks, that makes sense and I do prefer to keep the business logic separate. It is probably because of the patterns I used previously (inheritance and top to bottom approach) that throw me off. I will try to refactor this.
